lib/mobility-core/src/Kernel/Utils/SlidingWindowCounters.hs

```diff
@@ -245,26 +245,79 @@ decrementByValueImpl mbTimeStamp val keyModifier key SlidingWindowOptions {..} =
 --    ** counts = {<TIMEBASED_KEY_FOR_POSITIVE_CASE>: positiveCases , <TIMEBASED_KEY_FOR_THE_TOTAL_CASES>: totalCases}
 
 -- | getLatestRatio :: (id to getResult for, and generate TIMEBASED_KEY_FOR_THE_TOTAL_CASES) -> (id modifier to create TIMEBASED_KEY_FOR_POSITIVE_CASE) -> Resultsant Ratio of the sliding window
-getLatestRatio ::
+-- Minutes | Hours | Days | Months | Years
+calculateRatioForFirstTime ::
   ( L.MonadFlow m,
     Redis.HedisFlow m r
   ) =>
   Text ->
   (Text -> Text) ->
   (Text -> Text) ->
+  (Text -> Text -> Text) ->
+  Text ->
   SlidingWindowOptions ->
   m Double
-getLatestRatio driverId mkPostiveCaseKeyfn mkTotalCaseKeyfn s@SlidingWindowOptions {..} = do
+calculateRatioForFirstTime driverId mkPostiveCaseKeyfn mkTotalCaseKeyfn mkOldRatioKeyfn ratioType s@SlidingWindowOptions {..} = do
   utcTime <- L.runIO getCurrentTime
   let positiveCaseKeysList = getkeysForLastPeriods s utcTime $ makeSlidingWindowKey periodType (mkPostiveCaseKeyfn driverId)
-  let totalCountKeysList = getkeysForLastPeriods s utcTime $ makeSlidingWindowKey periodType (mkTotalCaseKeyfn driverId)
+      totalCountKeysList = getkeysForLastPeriods s utcTime $ makeSlidingWindowKey periodType (mkTotalCaseKeyfn driverId)
   positiveCases <- sum <$> mapMaybeM Redis.get positiveCaseKeysList
   totalCases <- nonZero . sum <$> mapMaybeM Redis.get totalCountKeysList
-  pure $ positiveCases / totalCases
+  let ratio = positiveCases / totalCases
+      expTime = fromInteger . floor $ diffUTCTime (incrementPeriod periodType utcTime) utcTime
+  Redis.setExp (mkOldRatioKeyfn driverId ratioType) ratio expTime
+  pure ratio
   where
     nonZero :: Double -> Double
     nonZero a = if a == 0.0 then 1.0 else a
 
+recalculateRatio ::
+  ( L.MonadFlow m,
+    Redis.HedisFlow m r
+  ) =>
+  Text ->
+  (Text -> Text) ->
+  (Text -> Text) ->
+  (Text -> Text -> Text) ->
+  Text ->
+  SlidingWindowOptions ->
+  m Double
+recalculateRatio driverId mkPostiveCaseKeyfn mkTotalCaseKeyfn mkOldRatioKeyfn ratioType s@SlidingWindowOptions {..} = do
+  let positiveCaseKey = mkPostiveCaseKeyfn driverId
+      totalCaseKey = mkTotalCaseKeyfn driverId
```

Comment:
these are not the correct keys where increment functions are incrementing the counters. the key has time info as well in them, you need to use the `makeSlidingWindowKey` here with current time to generate correct keys.